## Install following dependencies/packages:

1. `transformers`
2. `langchain`
3. `langchain-huggingface`

For installing use following command in the google colab terminal:
`pip install <package>`

## Create & Import token from Hugging Face into Google Colab

1. Creating a token from hugging face:-

 Account Profile -> Access Tokens -> Create new token -> give token a name -> switch over to 'Read' tab -> Create token -> Copy token

2. Importing token into Google Colab:-

  Open terminal in Colab -> enter `huggingface-cli login` command -> It will ask you to paste the token (Note: After pasting, the token will NOT be visible) -> Press 'Y' for git credential -> You should see 'Current Active token: ___' with your token name.

## Google Colab Notebook

### 1. Importing packages and model

*Note: Make sure you have selected runtime with GPU in google colab, this will speed up the computing*

You can check if you have GPU in your runtime by using following code, it should return `True`

In [39]:
# importing PyTorch package and using it's 'cuda.is_available()' command to check if GPU is available
import torch
print(torch.cuda.is_available())

True


In [40]:
# importing necessory packages
from transformers import pipeline

# importing/selecting model
model = pipeline(task="summarization", model="facebook/bart-large-cnn")

Device set to use cuda:0


### 2. Getting text from user


For this testing exmaple I'm using this latest news article from the verge: https://www.theverge.com/news/707106/cyberpunk-2077-mac-release-date


**Full Article:**

Cyberpunk 2077 is coming to macOS later this week, after it was originally set to debut in early 2025. Cyberpunk 2077: Ultimate Edition will officially be available for Apple silicon devices for the first time on July 17th, from Apple’s own Mac App Store, Steam, Epic Games Store, and GOG.com.

CD Projekt Red has optimized Cyberpunk 2077 for a variety of Mac hardware, all the way from the early M1 models up to the latest M4 chip family. Only some M1 and M2 Macs are supported, though — you’ll need a Mac with 16GB or more of unified memory to play Cyberpunk 2077, so base 8GB M1 and M2 chips won’t be compatible.

There will be graphics presets that are optimized for each Apple silicon model, and more powerful Apple silicon chips will also fully support features like path tracing, which is a film-like full ray tracing mode that greatly improves lighting and occlusion in the game.

AMD’s FSR upscaling and frame generation will also be available in Cyberpunk 2077 on Mac, helping the game feel more smooth with options like path tracing enabled. There will be MetalFX upscaling, too, alongside support for Apple’s Magic Mouse and Trackpad hardware.

Apple and CD Projekt Red haven’t shared performance data for what to expect across different Apple silicon Mac configurations. If you own an Apple XDR display, or any other HDR monitor, you’ll get HDR support in Cyberpunk 2077, and there’s even spatial audio and head-tracking with Apple AirPods.

Mac owners have technically been able to run Cyberpunk 2077 on their devices for a couple years now, thanks to Apple’s own Proton-like game porting tool. This official release means the game takes full advantage of Apple silicon and Apple’s native Metal graphics APIs, including its Tile-Based Deferred Rendering (TBDR) architecture.

Cyberpunk 2077 debuts on Mac just as CD Projekt Red releases the latest 2.3 patch for the game, which includes additional cars and new photo mode features. The game will support cross progression across all platforms and includes the Phantom Liberty expansion and all of the major updates to the game so far.


In [41]:
input_text = input('Please enter the text you would like to summarize: ')

Please enter the text you would like to summarize: Cyberpunk 2077 is coming to macOS later this week, after it was originally set to debut in early 2025. Cyberpunk 2077: Ultimate Edition will officially be available for Apple silicon devices for the first time on July 17th, from Apple’s own Mac App Store, Steam, Epic Games Store, and GOG.com.  CD Projekt Red has optimized Cyberpunk 2077 for a variety of Mac hardware, all the way from the early M1 models up to the latest M4 chip family. Only some M1 and M2 Macs are supported, though — you’ll need a Mac with 16GB or more of unified memory to play Cyberpunk 2077, so base 8GB M1 and M2 chips won’t be compatible.  There will be graphics presets that are optimized for each Apple silicon model, and more powerful Apple silicon chips will also fully support features like path tracing, which is a film-like full ray tracing mode that greatly improves lighting and occlusion in the game.  AMD’s FSR upscaling and frame generation will also be availa

#### Calculating input tokens:

In [42]:
input_tokens = len(model.tokenizer(input_text)[0])        # can also use 'input_ids' instead of [0]th position
print(f'The text entered generated {input_tokens} tokens and they will be used as an input to the model')

The text entered generated 491 tokens and they will be used as input to the model


### 3. Model output

In [46]:
# model output
response = model(input_text)

# displaying model output
print('Output Summary:- ')
print(response[0]['summary_text'])

Output Summary:- 
Cyberpunk 2077: Ultimate Edition will be available for Apple silicon devices for the first time on July 17th. CD Projekt Red has optimized the game for a variety of Mac hardware, from the early M1 models up to the latest M4 chip family. Only some M1 and M2 Macs are supported, though — you’ll need a Mac with 16GB or more of unified memory.


#### Calculating output tokens


In [47]:
summary = response[0]['summary_text']
output_tokens = len(model.tokenizer(summary)[0])
print(f'Output Tokens = {output_tokens}')

Output Tokens = 88


### 4. Fine tuning the model

#### If we consider another example with shorter text as an input

In [55]:
# input
text = '''Renewable energy is an essential component of sustainable development. Unlike fossil fuels, which are finite and contribute to environmental degradation, renewable energy sources such as solar, wind, and hydroelectric power offer a clean and sustainable alternative. These energy sources are abundant and can significantly reduce greenhouse gas emissions, helping to combat climate change. By investing in renewable energy, countries can create jobs, enhance energy security, and promote environmental sustainability.'''

# calculating input tokens
input_tokens = len(model.tokenizer(text)[0])        # can also use 'input_ids' instead of [0]th position
print(f'Input tokens: {input_tokens}')

Input tokens: 86


In [60]:
# selecting the model
model = pipeline(task="summarization", model="facebook/bart-large-cnn")

# model output
response = model(text)

print()
# displaying model output
print('Output Summary:- ')
print(response[0]['summary_text'])

print()
# calculating output tokens
summary = response[0]['summary_text']
output_tokens = len(model.tokenizer(summary)[0])
print(f'Output Tokens = {output_tokens}')

Device set to use cuda:0
Your max_length is set to 142, but your input_length is only 85. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=42)



Output Summary:- 
Renewable energy is an essential component of sustainable development. Unlike fossil fuels, renewable energy sources such as solar, wind, and hydroelectric power offer a clean and sustainable alternative. These energy sources are abundant and can significantly reduce greenhouse gas emissions. By investing in renewable energy, countries can create jobs, enhance energy security, and promote environmental sustainability.

Output Tokens = 71


In the output, the 2nd line is telling us that the length of our input text is too short beacuse it only contained 86 tokens and the model expected a lot more.


By default the `max_length`, length of output tokens for this model, is set to 142 i.e. if you provide less than or close to 142 tokens as an input then basically the summarization task might not work and model will just give us rephrased ouput.

We can see just that by comparing length of input & output tokens, that model just rephrased the text that was given as an input

#### Making changes to the model i.e. Fine-Tuning

Since this is a summarization problem, the ouput of the model needs to be less than input.

I'm gonna set the output tokens to be 50% of the input tokens

In [57]:
# Adjusting the output to be 50% of the input
desired_max_length = int(input_tokens * 0.5)

# if above calculated value is less than 25, then set it to 25. This will ensure user-understable output for really short passages.
if desired_max_length < 25:
  desired_max_length = 25

In [58]:
# fine tuning the model
response = model(text, max_length=desired_max_length, min_length=5, do_sample=False)

`max_length`: calculated above

`min_length`: The minimum length of the generated summary in tokens. This helps ensure the summary isn't too short to be useful.

`do_sample`: Set to 'False' for deterministic output (recommended for most summarization tasks unless you want creative variations).

In [59]:
# Output of the model
print('Output Summary:- ')
print(response[0]['summary_text'])

print()
# Calculating output tokens
summary = response[0]['summary_text']
output_tokens = len(model.tokenizer(summary)[0])
print(f'Output Tokens = {output_tokens}')

Output Summary:- 
Renewable energy is an essential component of sustainable development. Unlike fossil fuels, renewable energy sources offer a clean and sustainable alternative. By investing in renewable energy, countries can create jobs, enhance energy security

Output Tokens = 42


Documentation about `pipeline()` and `max_length`,`min_length` & `do_sample` can be found at:

https://huggingface.co/docs/transformers/main_classes/pipelines

https://huggingface.co/docs/transformers/main_classes/text_generation